aten/src/ATen/core/DistributionsHelper.h

```diff
@@ -0,0 +1,298 @@
+ */
+template <typename T>
+struct normal_distribution {
+
+  C10_HOST_DEVICE inline normal_distribution(T mean_in, T stdv_in) {
+    #ifndef __HIP_PLATFORM_HCC__
+      #ifdef __CUDA_ARCH__
+        assert(stdv_in > 0);
+      #else
+        AT_ASSERT(stdv_in > 0);
+      #endif
+    #endif
+    mean = mean_in;
+    stdv = stdv_in;
+  }
+
+  template<typename U = T> 
+  C10_HOST inline typename std::enable_if<std::is_same<U, double>::value, DOUBLE2>::type
+  operator()(at::CPUGenerator* generator){
+    uniform_real_distribution<double> uniform(0.0, 1.0);
+    DOUBLE2 result;
+    double u2 = uniform(generator);
+    double u1 = uniform(generator);
+    // extra pre-caution to make sure log never gets zero
+    double log_val = 1.0 - u1;
+    if (log_val == 0.0) {
+      log_val = std::numeric_limits<double>::min();
+    }
+    double r = ::sqrt(-2.0 * ::log(log_val));
+    result[0] = r * ::cos(2.0 * M_PI * u2) * stdv + mean;
+    result[1] = r * ::sin(2.0 * M_PI * u2) * stdv + mean;
+    return result;
+  }
+
+  template<typename U = T> 
+  C10_HOST inline typename std::enable_if<std::is_same<U, float>::value
+                                          || std::is_same<U, at::Half>::value, FLOAT2>::type
+  operator()(at::CPUGenerator* generator){
+    uniform_real_distribution<float> uniform(0.0, 1.0);
+    FLOAT2 result;
+    float u2 = uniform(generator);
+    float u1 = uniform(generator);
+    // extra pre-caution to make sure log never gets zero
+    float log_val = 1.0 - u1;
+    if (log_val == 0.0f) {
+      log_val = std::numeric_limits<float>::min();
+    }
+    float r = ::sqrt(-2.0 * ::log(log_val));
+    result[0] = r * ::cos(2.0 * M_PI * u2) * stdv + mean;
+    result[1] = r * ::sin(2.0 * M_PI * u2) * stdv + mean;
+    return result;
+  }
+
+  private:
+    T mean;
+    T stdv;
+};
+
+/**
+ * Samples a bernoulli distribution given a probability input
+ */
+template <typename T>
+struct bernoulli_distribution {
+
+  C10_HOST_DEVICE inline bernoulli_distribution(T p_in) {
+    #ifndef __HIP_PLATFORM_HCC__
+      #ifdef __CUDA_ARCH__
+        assert(p_in >= 0 && p_in <= 1);
+      #else
+        AT_ASSERT(p_in >= 0 && p_in <= 1);
+      #endif
+    #endif
+    p = p_in;
+  }
+
+  C10_HOST inline T operator()(at::CPUGenerator* generator) { 
+    uniform_real_distribution<T> uniform(0.0, 1.0);
+    return uniform(generator) <= p;
+  }
+
+  private:
+    T p;
+};
+
+/**
+ * Samples a geometric distribution given a probability input
+ */
+template <typename T>
+struct geometric_distribution {
+
+  C10_HOST_DEVICE inline geometric_distribution(T p_in) {
+    #ifndef __HIP_PLATFORM_HCC__
+      #ifdef __CUDA_ARCH__
+        assert(p_in > 0 && p_in < 1);
+      #else
+        AT_ASSERT(p_in > 0 && p_in < 1);
+      #endif
+    #endif
+    p = p_in;
+  }
+
+  C10_HOST inline int operator()(at::CPUGenerator* generator) {
+    uniform_real_distribution<T> uniform(0.0, 1.0);
+    auto sample = 1 - uniform(generator);
```

Comment:
Two lines below we use
static_cast<T>(0.0)
but here just "1".
Pehaps, more readable would be to use
T(1)
T(0)